File: MobileAcceleratorExample/URLConnectionVC.m
//
//  URLConnectionVC.m
//  MobileAcceleratorExample
//
//  Created by Brian Salomon on 11/13/15.
//  Copyright © 2015 Akamai. All rights reserved.
//

#import "URLConnectionVC.h"

@interface URLConnectionVC ()

@property (nonatomic, strong) UIImage *image;
@property (nonatomic, strong) NSMutableData *responseData;
@property (nonatomic, assign) NSInteger responseStatusCode;
@property (nonatomic) NSTimeInterval startTime;
@property (nonatomic) NSTimeInterval loadTimeElapsed;

@end

@implementation URLConnectionVC

- (void)viewDidLoad {
    [super viewDidLoad];

    self.startTime = [NSDate timeIntervalSinceReferenceDate];

    // disable NSURLConnection cache to demonstrate SDK
    NSURLRequest *request = [NSURLRequest requestWithURL:self.url];

    // Demonstrate using deprecated NSURLConnection
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
    NSURLConnection *connection = [[NSURLConnection alloc] initWithRequest:request delegate:self];
#pragma GCC diagnostic pop

    if (! connection) {
        NSLog(@"Error creating NSURLConnection.");
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

// run on main thread only due to UIKit reference
- (void) updateImage
{
    self.image = [UIImage imageWithData:self.responseData];
    [self.imageView setImage:self.image];
}

#pragma mark NSURLConnection callbacks

- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response
{
    // called on response or redirect
    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
    if (httpResponse) {
        self.responseStatusCode = httpResponse.statusCode;
    }

    self.responseData = [[NSMutableData alloc] init];
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data
{
    [self.responseData appendData:data];
}

- (NSCachedURLResponse *)connection:(NSURLConnection *)connection
                  willCacheResponse:(NSCachedURLResponse*)cachedResponse
{
    // return nil to disable the shared cache
    return cachedResponse;
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection
{
    NSTimeInterval timeNow = [NSDate timeIntervalSinceReferenceDate];
    self.loadTimeElapsed = timeNow - self.startTime;

    [self.connectionDelegate loadedConnectionSiteName:self.siteName];

    dispatch_async(dispatch_get_main_queue(), ^{
        [self updateImage];

        self.statusLabel.text = [NSString stringWithFormat:@"Status Code: %ld\nSize: %ld bytes",
                                 (long)self.responseStatusCode, (unsigned long)self.responseData.length];
    });
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error
{
    NSLog(@"NSURLConnection Error: %@", error);
    dispatch_async(dispatch_get_main_queue(), ^{
        self.statusLabel.text = [error localizedDescription];
    });
}

@end

File: MobileAcceleratorExampleSwift/URLConnectionVC.swift
//
//  URLConnectionVC.swift
//  MobileAcceleratorExampleSwift
//
//  Created by Waazim Reza on 23/05/2016.
//  Copyright © 2016 Akamai. All rights reserved.
//

import Foundation

protocol ConnectionVCDelegate: class {
	func loadedConnectionSiteName(siteName: String)
}

final class URLConnectionVC: UIViewController, NSURLConnectionDelegate, NSURLConnectionDataDelegate {

	var siteName: String = ""
	var url: URL?
	weak var connectionDelegate: ConnectionVCDelegate?

	@IBOutlet weak var imageView: UIImageView!
	@IBOutlet weak var statusLabel: UILabel!

	private var image: UIImage!
	private var responseData: NSMutableData!
	private var responseStatusCode: NSInteger!

	private var startTime: TimeInterval = 0.0
	private var loadTimeElapsed: TimeInterval = 0.0

	override func viewDidLoad() {
		super.viewDidLoad()

		guard let url = url else {
			print("No URL set")
			return
		}

		startTime = Date.timeIntervalSinceReferenceDate

		let request = URLRequest.init(url: url)

		// Demonstrate using deprecated NSURLConnection
		_ = NSURLConnection.init(request: request, delegate: self)
	}

	// run on main thread due to UIKit reference
	func updateImage() {
		if let responseData = responseData as Data? {
			image = UIImage.init(data: responseData)
			imageView.image = image
		}
	}

	// MARK: NSURLConnection callbacks

	func connection(_ connection: NSURLConnection, didReceive response: URLResponse) {
		// called on response or redirect
		if let httpResponse = response as? HTTPURLResponse {
			responseStatusCode = httpResponse.statusCode
		}

		responseData = NSMutableData()
	}

	func connection(_ connection: NSURLConnection, didReceive data: Data) {
		responseData.append(data)
	}

	func connection(_ connection: NSURLConnection, willCacheResponse cachedResponse: CachedURLResponse) -> CachedURLResponse? {
		// return nil to disable the shared cache
		return cachedResponse
	}

	func connectionDidFinishLoading(_ connection: NSURLConnection) {
		let timeNow: TimeInterval = Date.timeIntervalSinceReferenceDate
		loadTimeElapsed = timeNow - startTime

		connectionDelegate!.loadedConnectionSiteName(siteName: self.siteName)

		DispatchQueue.main.async() {
			self.updateImage()
			self.statusLabel.text = String(format: "Status Code: %ld\nSize: %ld bytes", self.responseStatusCode, self.responseData.length)
		}
	}

	func connection(_ connection: NSURLConnection, didFailWithError error: Error) {
		print("NSURLConnection Error: \(error.localizedDescription)")
        DispatchQueue.main.async() {
            self.statusLabel.text = error.localizedDescription
        }
	}
}
